File: src/main/java/com/blogly/blogly/application/post/GetPostByIdUseCase.java
package com.blogly.blogly.application.post;

import com.blogly.blogly.application.post.dto.PostDetailsResponse;
import com.blogly.blogly.domain.post.PostId;
import com.blogly.blogly.domain.post.PostNotFoundException;
import com.blogly.blogly.domain.post.PostRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@RequiredArgsConstructor
@Component
public class GetPostByIdUseCase {

    private final PostRepository repository;

    public PostDetailsResponse execute(PostId id) {
        return repository.findById(id)
                .map(PostDetailsResponse::from)
                .orElseThrow(() -> new PostNotFoundException(id));
    }
}

File: src/main/java/com/blogly/blogly/application/post/GetPostByIdUseCase.kt
package com.blogly.blogly.application.post

import com.blogly.blogly.application.post.dto.PostDetailsResponse
import com.blogly.blogly.application.post.dto.toDetailsResponse
import com.blogly.blogly.application.shared.UserProvider
import com.blogly.blogly.domain.post.PostId
import com.blogly.blogly.domain.post.PostRepository
import com.blogly.blogly.domain.post.exception.PostNotFoundException
import com.blogly.blogly.domain.shared.domainCheck
import org.springframework.stereotype.Component

@Component
class GetPostByIdUseCase(
    private val repository: PostRepository,
    private val userProvider: UserProvider
) {
    fun execute(postId: PostId): PostDetailsResponse {
        val post = repository.findById(postId) ?: throw PostNotFoundException(postId)
        domainCheck(post.isVisibleTo(userProvider.currentUserIdOrNull())) { PostNotFoundException(postId) }

        return post.toDetailsResponse()
    }
}
